File: packages/guardiandb-odm/src/schema.ts
import { ValidationError } from "./errors.js";
import { assertDocument, clone, randomId } from "./utils.js";
import type {
  Document,
  FieldSchema,
  NormalizedFieldSchema,
  NormalizedSchema,
  ScalarFieldType,
  SchemaDefinition,
} from "./types.js";

export function defineSchema<T extends Document>(definition: SchemaDefinition<T>): SchemaDefinition<T> {
  return definition;
}

export function normalizeSchema<T extends Document>(definition?: SchemaDefinition<T>): NormalizedSchema {
  const fields: Record<string, NormalizedFieldSchema> = {};
  let primaryKey: string | undefined;

  for (const [name, field] of Object.entries(definition?.fields ?? {})) {
    const normalized = normalizeField(field);
    if (normalized.primaryKey) {
      if (primaryKey !== undefined) {
        throw new ValidationError(name, "only one primary key is supported");
      }
      primaryKey = name;
      normalized.required = true;
      normalized.unique = true;
      normalized.index = true;
    }
    if (normalized.unique) {
      normalized.index = true;
    }
    fields[name] = normalized;
  }

  primaryKey ??= Object.hasOwn(fields, "_id") ? "_id" : Object.hasOwn(fields, "id") ? "id" : "_id";
  if (!Object.hasOwn(fields, primaryKey)) {
    fields[primaryKey] = {
      type: "string",
      required: true,
      nullable: false,
      primaryKey: true,
      unique: true,
      index: true,
    };
  } else {
    fields[primaryKey]!.primaryKey = true;
    fields[primaryKey]!.required = true;
    fields[primaryKey]!.unique = true;
    fields[primaryKey]!.index = true;
  }

  const timestamps = normalizeTimestamps(definition?.timestamps);
  if (timestamps !== undefined) {
    fields[timestamps.createdAt] ??= normalizeField({ type: "timestamp" });
    fields[timestamps.updatedAt] ??= normalizeField({ type: "timestamp" });
  }

  return {
    fields,
    primaryKey,
    strict: definition?.strict ?? definition !== undefined,
    ...(timestamps === undefined ? {} : { timestamps }),
    version: definition?.version ?? 1,
  };
}

export function prepareInsert(schema: NormalizedSchema, input: Document): Document {
  const document = clone(input);
  assertDocument(document);

  for (const [field, definition] of Object.entries(schema.fields)) {
    if (document[field] === undefined && definition.default !== undefined) {
      document[field] = clone(
        typeof definition.default === "function"
          ? (definition.default as () => unknown)()
          : definition.default,
      );
    }
  }

  if (document[schema.primaryKey] === undefined || document[schema.primaryKey] === null) {
    document[schema.primaryKey] = randomId();
  }

  if (schema.timestamps !== undefined) {
    const now = new Date().toISOString();
    document[schema.timestamps.createdAt] ??= now;
    document[schema.timestamps.updatedAt] = now;
  }

  validateDocument(schema, document);
  return document;
}

export function touchUpdatedAt(schema: NormalizedSchema, document: Document): void {
  if (schema.timestamps !== undefined) {
    document[schema.timestamps.updatedAt] = new Date().toISOString();
  }
}

export function validateDocument(schema: NormalizedSchema, document: Document): void {
  assertDocument(document);

  for (const [field, definition] of Object.entries(schema.fields)) {
    const value = document[field];
    if (value === undefined) {
      if (definition.required) {
        throw new ValidationError(field, "required field is missing");
      }
      continue;
    }
    if (value === null) {
      if (definition.required && !definition.nullable) {
        throw new ValidationError(field, "required field cannot be null");
      }
      continue;
    }
    if (!matchesType(definition.type, value)) {
      throw new ValidationError(field, `expected ${definition.type}`);
    }
    const customResult = definition.validate?.(value, document);
    if (customResult === false) {
      throw new ValidationError(field, "custom validator rejected the value");
    }
    if (typeof customResult === "string") {
      throw new ValidationError(field, customResult);
    }
  }

  if (schema.strict) {
    for (const field of Object.keys(document)) {
      if (!Object.hasOwn(schema.fields, field)) {
        throw new ValidationError(field, "field is not declared in the strict schema");
      }
    }
  }
}

function normalizeField(field: FieldSchema = {}): NormalizedFieldSchema {
  return {
    type: normalizeType(field.type),
    required: field.required ?? false,
    nullable: field.nullable ?? false,
    primaryKey: field.primaryKey ?? false,
    unique: field.unique ?? false,
    index: field.index ?? false,
    ...(field.default === undefined ? {} : { default: field.default }),
    ...(field.validate === undefined
      ? {}
      : { validate: field.validate as (value: unknown, document: Readonly<Document>) => boolean | string }),
  };
}

function normalizeType(type: FieldSchema["type"]): ScalarFieldType {
  if (type === undefined) return "any";
  if (type === String) return "string";
  if (type === Number) return "number";
  if (type === Boolean) return "boolean";
  if (type === Object) return "object";
  if (type === Array) return "array";
  if (typeof type === "string") return type;
  return "any";
}

function normalizeTimestamps(
  timestamps: SchemaDefinition["timestamps"],
): { createdAt: string; updatedAt: string } | undefined {
  if (timestamps === undefined || timestamps === false) {
    return undefined;
  }
  if (timestamps === true) {
    return { createdAt: "createdAt", updatedAt: "updatedAt" };
  }
  return {
    createdAt: timestamps.createdAt ?? "createdAt",
    updatedAt: timestamps.updatedAt ?? "updatedAt",
  };
}

function matchesType(type: ScalarFieldType, value: unknown): boolean {
  switch (type) {
    case "any":
      return true;
    case "string":
    case "timestamp":
      return typeof value === "string";
    case "number":
      return typeof value === "number" && Number.isFinite(value);
    case "boolean":
      return typeof value === "boolean";
    case "object":
      return value !== null && typeof value === "object" && !Array.isArray(value);
    case "array":
      return Array.isArray(value);
  }
}

File: src/odm/schema.rs
use crate::odm::error::{OdmError, Result};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::collections::{BTreeMap, BTreeSet};

/// Runtime field types understood by the ODM validator.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FieldType {
    Any,
    String,
    Number,
    Boolean,
    Object,
    Array,
    Timestamp,
}

impl FieldType {
    /// Returns whether a JSON value satisfies this field type (`Any` matches all).
    pub fn matches(self, value: &Value) -> bool {
        match self {
            Self::Any => true,
            Self::String | Self::Timestamp => value.is_string(),
            Self::Number => value.is_number(),
            Self::Boolean => value.is_boolean(),
            Self::Object => value.is_object(),
            Self::Array => value.is_array(),
        }
    }
}

/// Declarative constraints for one document field.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FieldDefinition {
    pub name: String,
    pub field_type: FieldType,
    pub required: bool,
    pub nullable: bool,
    pub primary_key: bool,
    pub unique: bool,
    pub indexed: bool,
}

impl FieldDefinition {
    /// Creates an unconstrained field definition of the given type.
    pub fn new(name: impl Into<String>, field_type: FieldType) -> Self {
        Self {
            name: name.into(),
            field_type,
            required: false,
            nullable: false,
            primary_key: false,
            unique: false,
            indexed: false,
        }
    }

    /// Marks the field as required (must be present).
    pub fn required(mut self) -> Self {
        self.required = true;
        self
    }

    /// Allows the field to hold an explicit `null` value.
    pub fn nullable(mut self) -> Self {
        self.nullable = true;
        self
    }

    /// Marks the field as the primary key (implies unique, indexed and required).
    pub fn primary_key(mut self) -> Self {
        self.primary_key = true;
        self.unique = true;
        self.indexed = true;
        self.required = true;
        self
    }

    /// Marks the field as unique (implies indexed).
    pub fn unique(mut self) -> Self {
        self.unique = true;
        self.indexed = true;
        self
    }

    /// Marks the field as indexed for faster queries.
    pub fn indexed(mut self) -> Self {
        self.indexed = true;
        self
    }
}

/// Names of the fields used to store creation and update timestamps.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct TimestampDefinition {
    pub created_at: String,
    pub updated_at: String,
}

/// Runtime representation of a typed or schemaless collection model.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ModelSchema {
    model_name: String,
    collection: String,
    fields: BTreeMap<String, FieldDefinition>,
    primary_key: String,
    strict: bool,
    auto_generate_primary_key: bool,
    timestamps: Option<TimestampDefinition>,
    version: u32,
}

impl ModelSchema {
    /// Creates a strict schema with the default `_id` primary key.
    pub fn new(model_name: impl Into<String>, collection: impl Into<String>) -> Self {
        Self {
            model_name: model_name.into(),
            collection: collection.into(),
            fields: BTreeMap::new(),
            primary_key: "_id".to_string(),
            strict: true,
            auto_generate_primary_key: true,
            timestamps: None,
            version: 1,
        }
    }

    /// A permissive schema with an automatically generated `_id` primary key.
    pub fn schemaless(collection: impl Into<String>) -> Self {
        let collection = collection.into();
        let mut schema = Self::new("Document", collection);
        schema.strict = false;
        schema.add_field(
            FieldDefinition::new("_id", FieldType::String)
                .primary_key()
                .required(),
        );
        schema
    }

    /// Adds a field definition (builder style) and returns the schema.
    pub fn field(mut self, field: FieldDefinition) -> Self {
        self.add_field(field);
        self
    }

    /// Adds or replaces a field definition, updating the primary key if the
    /// field is marked as one.
    pub fn add_field(&mut self, field: FieldDefinition) {
        if field.primary_key {
            self.primary_key = field.name.clone();
        }
        self.fields.insert(field.name.clone(), field);
    }

    /// Sets the collection name this schema maps to.
    pub fn set_collection(&mut self, collection: impl Into<String>) {
        self.collection = collection.into();
    }

    /// Enables or disables strict mode (rejecting undeclared fields).
    pub fn set_strict(&mut self, strict: bool) {
        self.strict = strict;
    }

    /// Builder-style variant of [`set_strict`](Self::set_strict).
    pub fn strict(mut self, strict: bool) -> Self {
        self.strict = strict;
        self
    }

    /// Enables or disables automatic primary-key generation.
    pub fn set_auto_generate_primary_key(&mut self, enabled: bool) {
        self.auto_generate_primary_key = enabled;
    }

    /// Builder-style variant of [`set_auto_generate_primary_key`](Self::set_auto_generate_primary_key).
    pub fn auto_generate_primary_key(mut self, enabled: bool) -> Self {
        self.auto_generate_primary_key = enabled;
        self
    }

    /// Sets the schema version.
    pub fn set_version(&mut self, version: u32) {
        self.version = version;
    }

    /// Enables timestamp fields, declaring them if they are not already present.
    pub fn enable_timestamps(
        &mut self,
        created_at: impl Into<String>,
        updated_at: impl Into<String>,
    ) {
        let created_at = created_at.into();
        let updated_at = updated_at.into();

        self.fields
            .entry(created_at.clone())
            .or_insert_with(|| FieldDefinition::new(created_at.clone(), FieldType::Timestamp));
        self.fields
            .entry(updated_at.clone())
            .or_insert_with(|| FieldDefinition::new(updated_at.clone(), FieldType::Timestamp));
        self.timestamps = Some(TimestampDefinition {
            created_at,
            updated_at,
        });
    }

    /// Builder-style variant of [`enable_timestamps`](Self::enable_timestamps).
    pub fn timestamps(
        mut self,
        created_at: impl Into<String>,
        updated_at: impl Into<String>,
    ) -> Self {
        self.enable_timestamps(created_at, updated_at);
        self
    }

    /// Returns the model name.
    pub fn model_name(&self) -> &str {
        &self.model_name
    }

    /// Returns the collection name this schema maps to.
    pub fn collection(&self) -> &str {
        &self.collection
    }

    /// Returns the declared field definitions keyed by name.
    pub fn fields(&self) -> &BTreeMap<String, FieldDefinition> {
        &self.fields
    }

    /// Returns the primary-key field name.
    pub fn primary_key(&self) -> &str {
        &self.primary_key
    }

    /// Returns whether strict mode is enabled.
    pub fn strict_mode(&self) -> bool {
        self.strict
    }

    /// Returns whether the primary key is auto-generated when missing.
    pub fn auto_generates_primary_key(&self) -> bool {
        self.auto_generate_primary_key
    }

    /// Returns the timestamp field definition, if timestamps are enabled.
    pub fn timestamp_definition(&self) -> Option<&TimestampDefinition> {
        self.timestamps.as_ref()
    }

    /// Returns the schema version.
    pub fn version(&self) -> u32 {
        self.version
    }

    /// Returns the set of field names that are indexed (including unique and
    /// primary-key fields).
    pub fn indexed_fields(&self) -> BTreeSet<String> {
        self.fields
            .values()
            .filter(|field| field.indexed || field.unique || field.primary_key)
            .map(|field| field.name.clone())
            .collect()
    }

    /// Returns the set of field names that must be unique (including the primary
    /// key).
    pub fn unique_fields(&self) -> BTreeSet<String> {
        self.fields
            .values()
            .filter(|field| field.unique || field.primary_key)
            .map(|field| field.name.clone())
            .collect()
    }

    /// Validates the schema itself: non-empty collection name, at most one
    /// primary key, and consistent primary-key metadata.
    pub fn validate_definition(&self) -> Result<()> {
        if self.collection.trim().is_empty() {
            return Err(OdmError::InvalidSchema(
                "collection name cannot be empty".to_string(),
            ));
        }

        let primary_fields: Vec<&FieldDefinition> = self
            .fields
            .values()
            .filter(|field| field.primary_key)
            .collect();
        if primary_fields.len() > 1 {
            return Err(OdmError::InvalidSchema(
                "only one primary key is supported".to_string(),
            ));
        }

        if let Some(primary) = primary_fields.first()
            && primary.name != self.primary_key
        {
            return Err(OdmError::InvalidSchema(format!(
                "primary key metadata is inconsistent for `{}`",
                primary.name
            )));
        }

        if self.strict && !self.fields.contains_key(&self.primary_key) {
            return Err(OdmError::InvalidSchema(format!(
                "primary key `{}` is not declared as a field",
                self.primary_key
            )));
        }

        Ok(())
    }

    /// Validates a document against the schema: required/nullable rules, field
    /// types, and (in strict mode) rejection of undeclared fields.
    pub fn validate_document(&self, document: &Value) -> Result<()> {
        let object = document.as_object().ok_or_else(|| OdmError::Validation {
            field: "$document".to_string(),
            message: "document must be a JSON object".to_string(),
        })?;

        for field in self.fields.values() {
            match object.get(&field.name) {
                None if field.required => {
                    return Err(OdmError::Validation {
                        field: field.name.clone(),
                        message: "required field is missing".to_string(),
                    });
                }
                Some(Value::Null) if field.required && !field.nullable => {
                    return Err(OdmError::Validation {
                        field: field.name.clone(),
                        message: "required field cannot be null".to_string(),
                    });
                }
                Some(Value::Null) => {}
                Some(value) if !field.field_type.matches(value) => {
                    return Err(OdmError::Validation {
                        field: field.name.clone(),
                        message: format!("expected {:?}", field.field_type),
                    });
                }
                _ => {}
            }
        }

        if self.strict {
            for key in object.keys() {
                // `_id` is an internal storage key when a model uses a custom primary key.
                if key != "_id" && !self.fields.contains_key(key) {
                    return Err(OdmError::Validation {
                        field: key.clone(),
                        message: "field is not declared in the strict schema".to_string(),
                    });
                }
            }
        }

        Ok(())
    }

    /// Returns the document as a mutable JSON object, erroring if it is not one.
    pub(crate) fn object_mut<'a>(
        &self,
        document: &'a mut Value,
    ) -> Result<&'a mut Map<String, Value>> {
        document
            .as_object_mut()
            .ok_or_else(|| OdmError::Validation {
                field: "$document".to_string(),
                message: "document must be a JSON object".to_string(),
            })
    }
}
